Adds setup command to bootstrap financial aid for DEDP

jkachel opened the pull request:
#### Pre-Flight checklist

- [X] Testing
  - [X] Code is tested
  - [X] Changes have been manually tested

#### What are the relevant tickets?

Fixes #883

#### What's this PR do?

Adds a management command that bootstraps DEDP financial aid. This includes:
- making sure a DEDP program exists
- creating discounts for the tiers (or identifying and reusing discounts that already exist, if their parameters are correct)
- creating tiers for DEDP financial assistance requests

The command will also deactivate (by setting the expiration date to a date in the past) discounts that follow the naming scheme (`DEDP-fa-tierX-YYYY` where YYYY is the year), and will deactivate tiers that aren't associated with a current set of DEDP discounts. 

#### How should this be manually tested?

Automated tests should pass.

See the comment block at the top of the command `flexiblepricing/management/commands/configure_for_dedp.py` for an explanation of what your database should have in it after running the command. 

To test full-stop bootstrapping, ensure you do not have any discounts that follow the `DEDP-fa-tier` naming scheme, nor tiers that use them, nor a DEDP program (`program-v1:MITx+DEDP`). Then, run the command. You should have a DEDP program, a set of four DEDP discounts, and a set of four tiers for the program that use the discounts. 

To test partial bootstrapping, run the command after first having any of these conditions:
- an existing DEDP program
- existing DEDP discounts for the current year
- tiers that use the discounts

The command should run successfully and the existing data should be kept if it matches the parameters. 

To test invalidation, run the command after first having any of these conditions:
- a DEDP discount for a prior year (so, for 2022, a discount with a code of `DEDP-fa-tierX-2021` or before)
- tiers that use the prior year discount

The discounts should be invalidated and the tiers should be turned off (`current=False`).

rachellougee commented on flexiblepricing/management/commands/configure_for_dedp.py:
```diff
@@ -0,0 +1,195 @@
+            discount_code__startswith="DEDP-fa-tier"
+        ).all()
+        self.stdout.write(
+            self.style.NOTICE(f"{len(discounts)} existing discounts for {current_year}")
```
Is this intended to assume existing discounts are for the current year?

rachellougee replied:
ignore me. I misread this text. I think my question was on the different line 135-138

rachellougee commented on flexiblepricing/management/commands/configure_for_dedp.py:
```diff
@@ -0,0 +1,195 @@
+                    found_discount = discount
+                    self.stdout.write(
+                        self.style.NOTICE(
+                            f"\tFound a discount ({discount.discount_code}"
```
Minor pick, missing )

rachellougee commented:
Overall it looks good. I was able to verify the inserted records after the initial run. 
I just have a question regarding the discounts. I manually updated the discount code - DEDP-fa-tier4-2022 to DEDP-fa-tier4-2021, then rerun the command. It didn't update the discount code to the correct year. By looking at the code, these didn't get updated because discounts are still valid (no expiration date), so it's assuming these are still valid even though the text said it's for 2021

```Setting up the program...
Using existing program 1
5 existing discounts for 2022
Looking for 750 dollars-off discount for threshold 0
        Found a discount (DEDP-fa-tier1-2021
        Updated existing tier ID 1 for threshold 0
Looking for 650 dollars-off discount for threshold 25000
        Found a discount (DEDP-fa-tier2-2021
        Updated existing tier ID 2 for threshold 25000
Looking for 500 dollars-off discount for threshold 50000
        Found a discount (DEDP-fa-tier3-2021
        Updated existing tier ID 3 for threshold 50000
Looking for 0 percent-off discount for threshold 75000
        Found a discount (DEDP-fa-tier4-2021
        Updated existing tier ID 4 for threshold 75000
1 discounts expired
0 tiers deactivated
```

jkachel commented:
Fixed the output stuff in https://github.com/mitodl/mitxonline/pull/898/commits/00fc4a77b2306be89eb4cc02ed77a6ec6617b778. Good catch on the discount code reuse - I updated the query to explicitly look for discounts for the current year and invalidate the ones that aren't (so if you have some from 2021, say, it will just turn those off). I thought about changing the codes but we don't track revisions of discount codes, so changing them would also change the redemptions for learners, and I don't think we also want to display the wrong year for folks who have been approved for flex pricing either.

rachellougee commented:
LGTM. Discount work as expected now
